### ISSUE TITLE
sql: drop table schema change can retry forever

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#26324]
Forked from https://github.com/cockroachdb/cockroach/issues/24029#issuecomment-392219788

> DistSender is presumably splitting this single operation into thousands of requests to the 10s of thousands of ranges in this table. Interestingly, this is not the only `ClearRange` operation logged from that line of code in the cluster:
>
```
I180525 23:24:02.838137 314 sql/tablewriter_delete.go:184  [n3] ClearRange /Table/51 - /Table/52
I180525 23:29:07.004769 303 sql/tablewriter_delete.go:184  [n10] ClearRange /Table/51 - /Table/52
I180525 23:34:11.599280 252 sql/tablewriter_delete.go:184  [n7] ClearRange /Table/51 - /Table/52
```

> Notice that the timestamps are ~5min apart. I believe this corresponds to the 5min `SchemaChangeLeaseDuration`. I haven't verified it yet, but I'd be willing to bet that the `ClearRange` operation is taking longer than 5min to complete which results in the schema change lease for this table being picked up by another node which then executes another `ClearRange` operation which will take longer than 5min and on and on. Why this eventually wedges the cluster is unclear.
>
> I'm going to try bumping `SchemaChangeLeaseDuration` to a much higher value to see if that allows the drop to succeed without wedging the cluster. I'm also going to turn on additional logging at the DistSender level to make sure that the `ClearRange` operation is making progress and not retrying RPCs for individual ranges over and over.

Subsequent testing did verify that the `ClearRange` operation is taking longer than 5min (this is the root issue of #24029) which results in the schema change lease expiring and getting acquired by another which performs the `ClearRange` operation again adding more range tombstones which horks up the cluster even more.

Two thoughts to address this problem:
* We could change the `ClearRange` operation to obey the `max_span_request_keys` parameter similar to `DeleteRange`. 
* We could pre-scan the range metadata for the span of keys being dropped and limit the `ClearRange` ops to a few hundred ranges at a time.

**Comment by tschottdorf** [comment]
How does this work for the slow delete path? It seems that the schema changer just has no notion of checkpointing longer-running jobs? (i.e. why is another node taking the lease if our node is actually doing work).

**Comment by petermattis** [comment]
`tableDeleter.legacyDeleteAllRowsFast` performs `DeleteRange` requests in 600 key chunks. The `ClearRange` path deletes the entire table with a single `ClearRange` request. I haven't piece together where checkpoint is done on the truncate path, but I assume there is a difference here.

**Comment by vivekmenezes** [comment]
the schema changer over the slow path deletes a chunk of keys and then hits the schema change lease so that it doesn't expire. There is no checkpoint preserved. The truncate path is the same as the drop path.

**Comment by vivekmenezes** [comment]
@benesch I know you've fixed the performance problems around DROP TABLE. Is this still an issue?

**Comment by benesch** [comment]
Nope, the specific issue here was fixed in ff25bec. There are more performance improvements coming in the next RocksDB bump, but those are tracked in other issues. Closing!
